container/state.go

```diff
@@ -271,3 +282,27 @@ func (s *State) SetDead() {
 	s.Dead = true
 	s.Unlock()
 }
+
+// SetCheckpointed sets the container's status to indicate it has been checkpointed
+func (s *State) SetCheckpointed(leaveRunning bool) {
+	s.Lock()
+	s.CheckpointedAt = time.Now().UTC()
+	s.Checkpointed = !leaveRunning
+	s.Running = leaveRunning
+	s.Paused = false
+	s.Restarting = false
+	// FIXME: Not sure if we need to close and recreate waitChan.
+	// close(s.waitChan)
+	// s.waitChan = make(chan struct{})
+	s.Unlock()
+}
+
+// HasBeenCheckpointed indicates whether the container has ever been checkpointed
+func (s *State) HasBeenCheckpointed() bool {
+	return s.CheckpointedAt != time.Time{}
+}
+
+// IsCheckpointed indicates whether the container is currently checkpointed
+func (s *State) IsCheckpointed() bool {
+	return s.Checkpointed
```

Comment:
In the original implementation in Docker 1.5 (https://github.com/SaiedKazemi/docker/tree/cr), I saw failures when Checkpointed wasn't protected by locks during concurrent tests (i.e., docker <cmd> on one terminal and docker checkpoint on another terminal).

[Sorry I've been away from this for a while, but Ross has been doing a great job rebasing and moving things forward.]